src/lib.rs

```diff
@@ -442,21 +473,88 @@ impl Log for Logger {
     }
 
     fn log(&self, record: &Record) {
-        let _ = match self.target {
-            Target::Stdout if self.matches(record) => (self.format)(&mut io::stdout(), record),
-            Target::Stderr if self.matches(record) => (self.format)(&mut io::stderr(), record),
-            _ => Ok(()),
-        };
+        if self.matches(record) {
+            // Log records are written to a thread-local buffer before being printed
+            // to the terminal. We clear these buffers afterwards, but they aren't shrinked
+            // so will always at least have capacity for the largest log record formatted
+            // on that thread.
+
+            thread_local! {
+                static FORMATTER: RefCell<Option<Formatter>> = RefCell::new(None);
+            }
+
+            FORMATTER.with(|tl_buf| {
+                let mut tl_buf = tl_buf.borrow_mut();
+
+                if tl_buf.is_none() {
+                    *tl_buf = Some(Formatter::new(self.writer.buffer()));
+                }
+
+                // The format is guaranteed to be `Some` by this point
+                let mut formatter = tl_buf.as_mut().unwrap();
+
+                let _ = (self.format)(&mut formatter, record)
+                    .and_then(|_| {
+                        let printed = self.writer.print(&formatter.buf);
+                        formatter.buf.clear();
+
+                        printed
+                    });
+            });
+        }
     }
 
     fn flush(&self) {}
 }
 
+impl Formatter {
+    fn new(buf: Buffer) -> Self {
+        Formatter {
+            buf: buf,
+        }
+    }
+
+    fn color(&mut self, color: Color) -> StyledFormatter<&mut Buffer> {
+        let mut spec = ColorSpec::new();
+        spec.set_fg(Some(color));
+
+        StyledFormatter {
+            buf: &mut self.buf,
+            spec: spec
+        }
+    }
+}
+
+impl Write for Formatter {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.buf.write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.buf.flush()
+    }
+}
+
+impl<W> Write for StyledFormatter<W>
+    where W: WriteColor
+{
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.buf.set_color(&self.spec)?;
+        let write = self.buf.write(buf);
+        self.buf.reset()?;
```

Comment:
Not returning the bytes written if the terminal fails to reset seems like a bad thing to do.